### PR TITLE
Initialize working dirs inside workspace dir

### LINKED ISSUE #725: Steps with workingDir refuse to start because the directory does not exist on pod startup
# Expected Behavior

We can set the workingDir to a subdirectory of the workspace directory (e.g. the one we plan to populate with a git repository).

# Actual Behavior

Pods fail because the containers can't `chdir` into a directory that does not exist. This is likely a bug due to the move from init containers (where the directory would have existed since they start in order).

# Steps to Reproduce the Problem

Create a task/taskrun with

1. a git input with targetPath "foo"
2. a step with workingDir "workspace/foo"

# Additional Info

### PR DESCRIPTION
**Pull request by dicarlo2**
# Changes

In order to allow steps to define workingDirs that are subdirectories of the workspace directory we need to make sure they have been created first since they will not exist on startup.

Fixes #725

Note: this PR does not solve the general problem of ensuring all valid combinations of workingDir + volumes + volume mounts will not fail, instead opting to keep things simple and solve the most common case as a first pass - setting the working directory to the cloned git repository (or other input resource folder). We also do not check that the `workingDir` path actually refers to the implicit workspace volume (as opposed to a user defined one that they've mounted at the same path) because it keeps the logic simple and the worst case is we have an additional empty directory in the implicit workspace volume.

Adds an additional parameter to the controller config for the image to use for this container. We could have reused the other bash image parameter, but it makes for a nice extension point to tekton to keep them separate IMO since it allows overriding specific bits of functionality should the user desire.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

```
Fix setting workingDirs that are subdirectories of the default workspace directory
```

### CODE REVIEW

File: config/controller.yaml [hunk 1]
```diff
@@ -35,6 +35,7 @@ spec:
           "-git-image", "github.com/tektoncd/pipeline/cmd/git-init",
           "-nop-image", "github.com/tektoncd/pipeline/cmd/nop",
           "-bash-noop-image", "github.com/tektoncd/pipeline/cmd/bash",
+          "-bash-working-dir-image", "github.com/tektoncd/pipeline/cmd/bash",
```
Comment (vdemeester) [inline config/controller.yaml]: Why not using `bash-noop-image` instead of having a new flag ? (we were also thinking of renaming this one as the `noop` make it more scarier than it is)

Comment (dicarlo2) [reply config/controller.yaml]: We could use the noop image, though my thinking was that it's nice to have separate arguments for users to override if necessary. I.e. if for some reason in a user's setup they need to override the image that's used for creating working directories, e.g. to have their own custom logic run, but don't want to change it for the current use of bash-noop-image (inside of the artifact setup). Thoughts?

Comment (dlorenc) [reply config/controller.yaml]: That SGTM @dicarlo2

**Comment by dlorenc** [comment]
This LGTM except for the question about the prefix check on WorkingDir

File: config/controller.yaml (hunk 1 above)
Comment (bobcatfish) [reply config/controller.yaml]: (besides @vdemeester I think we can actually get rid of the noop-image now that we're not using init containers right? i'm having deja vu like we already discussed this... :thinking::thinking::thinking:)

File: pkg/reconciler/v1alpha1/taskrun/resources/pod.go [hunk 2]
```diff
@@ -155,6 +161,39 @@ func makeCredentialInitializer(serviceAccountName, namespace string, kubeclient
 	}, volumes, nil
 }
 
+func makeWorkingDirInitializer(taskSpec v1alpha1.TaskSpec) *corev1.Container {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/resources/pod.go]: I'd really like to see this as a publicly exported function with its own unit tests, and i think we can tighten up the interface a bit more: e.g. pass in just the `steps` instead of the entire task spec (or do the work to extract the workingDirs from the `steps` before calling this function?)

then if we had that we could take or leave the additional cases in `TestMakePod` which i think are a bit more verbose than we'd see in some slightly more targeted unit tests

File: pkg/reconciler/v1alpha1/taskrun/resources/pod.go [hunk 3]
```diff
@@ -155,6 +161,39 @@ func makeCredentialInitializer(serviceAccountName, namespace string, kubeclient
...
+	for _, step := range taskSpec.Steps {
+		if step.WorkingDir != "" {
+			p := filepath.Clean(step.WorkingDir)
+			if strings.HasPrefix(p, workspaceDir) && p != workspaceDir && !seen[p] {
+				seen[p] = true
+				command := fmt.Sprintf("mkdir -p %s", p)
+				if script == "" {
+					script = command
+				} else {
+					script = fmt.Sprintf("%s && %s", script, command)
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/resources/pod.go]: id love to see a function that _only_ creates the `script` for a list of `workingDirs` with its own unit tests

File: config/controller.yaml (hunk 1 above)
Comment (bobcatfish) [reply config/controller.yaml]: (ahahahaa that was a different image and you already removed it @dicarlo2 :rofl: :rofl: :rofl: https://github.com/tektoncd/pipeline/pull/715 thank you :pray: ... meanwhile we have got to get `noop` out of the name of that flag so ppl like me dont get confuse...)

Comment (bobcatfish) [reply config/controller.yaml]: ... and #715 hasn't been merged yet ... i'm going to catch up eventually i promise...

**Comment by dicarlo2** [comment]
> This LGTM except for the question about the prefix check on WorkingDir

I might just be blind, but I'm not seeing the question. With that said, I did change the check to look at the relpath rather than the prefix, so maybe I fixed it :)

File: pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go [hunk 4]
```diff
@@ -249,6 +251,52 @@ func TestMakePod(t *testing.T) {
 			},
 			Volumes: implicitVolumes,
 		},
+	}, {
+		desc: "working-dir-in-workspace-dir",
```
Comment (dicarlo2) [inline pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go]: I left this test in so that there's at least one test sanity checking that we add the container as expected.

File: pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go [hunk 5]
```diff
@@ -309,3 +357,25 @@ func TestMakePod(t *testing.T) {
...
+	for _, c := range []struct {
+		desc        string
+		workingDirs []string
+		want        string
+	}{{
+		desc:        "simple",
+		workingDirs: []string{"/workspace/foo", "/workspace/bar", "/baz"},
+		want:        "mkdir -p /workspace/foo /workspace/bar",
+	}, {
+		desc:        "empty",
```
Comment (chmouel) [inline pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go]: nit: maybe add a test that checks for workingDirs and returns empty

Comment (dicarlo2) [reply pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go]: Sorry, I'm not following (maybe because I'm tired and it's pretty late here). Can you restate/give an example?

**Comment by dicarlo2** [comment]
> I also want to know, have you checked how much overhead this adds to TaskRuns that execute the extra container?

The container executes in < 1 second (GKE n1-standard-8, fwiw). I'd expect the overhead would be dominated by the time taken to download the container, which will vary across user setups. On that note, it might make sense to combine some of the binaries into a single image by default, in particular for the init container images (place tools, and this one).

File: pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go (hunk 5 above)
Comment (chmouel) [reply pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go]: something like this ? https://gist.github.com/chmouel/b1b085d96ffd4b7a620c26ed3708462c

Comment (dicarlo2) [reply pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go]: Done, thanks for clarifying 😄

**Comment by chmouel** [comment]
> On that note, it might make sense to combine some of the binaries into a single image by default, in particular for the init container images (place tools, and this one).

👍 that would be a great optimisation indeed!

**Comment by dicarlo2** [comment]
Oh, also, I expect the integration tests will fail since the `override-with-bash-working-dir:latest` image does not exist - if someone could give me a hand with fixing it that'd be awesome :)

File: pkg/reconciler/v1alpha1/taskrun/resources/pod.go [hunk 6]
```diff
@@ -91,9 +92,14 @@ const (
 	unnamedInitContainerPrefix = "build-step-unnamed-"
 	// Name of the credential initialization container.
 	credsInit = "credential-initializer"
+	// Name of the working dir initialization container.
+	workingDirInit = "working-dir-initializer"
 )
 
 var (
+	// The container used to initialize working directories before the build runs.
+	bashWorkingDirImage = flag.String("bash-working-dir-image", "override-with-bash-working-dir:latest",
```
Comment (vdemeester) [inline pkg/reconciler/v1alpha1/taskrun/resources/pod.go]: The image could refer `bash` image

```go
bashWorkingDirImage = flag.String("bash-working-dir-image", "override-with-bash-noop:latest", …
```

That way we don't need to publish a new image but it can be customized at runtime :angel:

File: pkg/reconciler/v1alpha1/taskrun/resources/pod.go [hunk 7]
```diff
@@ -160,6 +166,47 @@ func makeCredentialInitializer(serviceAccountName, namespace string, kubeclient
...
+				}
+			}
+		}
+	}
+
+	return script
+}
+
+func makeWorkingDirInitializer(steps []corev1.Container) *corev1.Container {
+	workingDirs := make([]string, len(steps))
```
Comment (vdemeester) [inline pkg/reconciler/v1alpha1/taskrun/resources/pod.go]: We could simplify this by making `workingDirs` a map (`map[string]bool` or even `map[string]struct{}`). That way we wouldn't have to worry about duplicates and thus wouldn't have to create the `seen` map above :angel:

**Comment by dlorenc** [comment]
> I might just be blind, but I'm not seeing the question. With that said, I did change the check to look at the relpath rather than the prefix, so maybe I fixed it :)

D'oh, yeah I think I didn't hit submit or something. Looks good now!

**Comment by dlorenc** [comment]
Looks like this needs a rebase and has a few comments left to address. @dicarlo2 are you still planning to get this fixed up?

**Comment by dicarlo2** [comment]
Yup, sorry about the delay, had another one of those weeks where I had too many things on my plate.

File: pkg/reconciler/v1alpha1/taskrun/resources/pod.go (hunk 6 above)
Comment (dicarlo2) [reply pkg/reconciler/v1alpha1/taskrun/resources/pod.go]: Done.

File: pkg/reconciler/v1alpha1/taskrun/resources/pod.go (hunk 7 above)
Comment (dicarlo2) [reply pkg/reconciler/v1alpha1/taskrun/resources/pod.go]: Good catch.

**Comment by dicarlo2** [comment]
Looks like the `override-with-bash-noop` image also does not exist.

**Comment by vdemeester** [comment]
#843 is merged, we can close this one 😁
/close
